### PR TITLE
Be less restritive when pruning redacted events

### PR DESCRIPTION
**Pull request by Florian14**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [ ] Technical
- [x] Other :

## Content

- Be less restrive when handling redacted events in the sdk
- Some cleanup using `StableUnstableId` object for event types with unstable ids
- Display debug notice for voice broadcast hidden events (when hidden events are shown in the timeline) 

## Motivation and context

Before, the sdk was only pruning events from a limited list, now the sdk will prune the events which are **not** in that list.

The custom events which were unknown from the sdk are now correctly pruned.   

Could be even less restrictive like Element-iOS but I decided to keep some state events / crypto events to limitate potential regressions

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt [hunk 1]
```diff
@@ -324,9 +325,11 @@ class MessageItemFactory @Inject constructor(
             informationData: MessageInformationData,
             highlight: Boolean,
             attributes: AbsMessageItem.Attributes
-    ): MessageVoiceItem? {
+    ): BaseEventItem<*>? {
         // Do not display voice broadcast messages
-        if (params.event.root.asMessageAudioEvent().isVoiceBroadcast()) return null
+        if (params.event.root.asMessageAudioEvent().isVoiceBroadcast()) {
+            return noticeItemFactory.create(params)
+        }
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt:332]: @Florian14 What is the purpose of this change here?

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/VoiceBroadcastItemFactory.kt [hunk 2]
```diff
@@ -47,16 +48,19 @@ class VoiceBroadcastItemFactory @Inject constructor(
...
             messageContent: MessageVoiceBroadcastInfoContent,
             highlight: Boolean,
             attributes: AbsMessageItem.Attributes,
-    ): AbsMessageVoiceBroadcastItem<*>? {
+    ): BaseEventItem<*>? {
         // Only display item of the initial event with updated data
-        if (messageContent.voiceBroadcastState != VoiceBroadcastState.STARTED) return null
+        if (messageContent.voiceBroadcastState != VoiceBroadcastState.STARTED) {
+            return noticeItemFactory.create(params)
+        }
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/VoiceBroadcastItemFactory.kt:63]: Same remark

**Review by giomfo — COMMENTED** [review]
commented - SGTM

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt (hunk 1 above)
Comment (Florian14) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt:332]: Display a notice if the event is hidden and the flag "show hidden events in the timeline" enabled in the labs

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/VoiceBroadcastItemFactory.kt (hunk 2 above)
Comment (Florian14) [reply vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/VoiceBroadcastItemFactory.kt:63]: same answer

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt [hunk 3]
```diff
@@ -61,45 +61,33 @@ internal class RedactionEventProcessor @Inject constructor() : EventInsertLivePr
...
-                EventType.MESSAGE,
-                in EventType.STATE_ROOM_BEACON_INFO,
-                in EventType.BEACON_LOCATION_DATA,
-                in EventType.POLL_START -> {
-                    Timber.d("REDACTION for message ${eventToPrune.eventId}")
-                    val unsignedData = EventMapper.map(eventToPrune).unsignedData
-                            ?: UnsignedData(null, null)
-
-                    // was this event a m.replace
+        when {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt:69]: Maybe add a `else` case to log a warning, which can help to detect possible issue  in the future.
Something like:

```kotlin
            else -> {
                Timber.w("Not pruning event (type $typeToPrune)")
            }
```

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt [hunk 4]
```diff
@@ -167,4 +155,34 @@ internal class RedactionEventProcessor @Inject constructor() : EventInsertLivePr
...
+                    eventType == EventType.STATE_ROOM_RELATED_GROUPS ||
+                    eventType == EventType.STATE_ROOM_PINNED_EVENT ||
+                    eventType == EventType.STATE_ROOM_ENCRYPTION ||
+                    eventType == EventType.STATE_ROOM_SERVER_ACL ||
+                    eventType == EventType.ROOM_KEY_REQUEST ||
+                    eventType == EventType.FORWARDED_ROOM_KEY ||
+                    eventType in EventType.ROOM_KEY_WITHHELD.values ||
+                    eventType == EventType.REQUEST_SECRET ||
+                    eventType == EventType.SEND_SECRET ||
+                    eventType == EventType.REACTION -> false
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt:184]: As discussed with @BillCarsonFr some event types in this list are only toDevice events, so should never be used here. The fact that all Event types are defined in `EventType` does not help.

Not blocking though.

**Review by bmarty — APPROVED** [review]
LGTM, thanks!

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt (hunk 3 above)
Comment (Florian14) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt:69]: Done https://github.com/vector-im/element-android/pull/7611/commits/f913b527501f39dfd2b0e53897d7b66585bb1d63

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt (hunk 4 above)
Comment (Florian14) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt:181]: I removed some to-device event types from the list, tell me if I am wrong https://github.com/vector-im/element-android/pull/7611/commits/57ef4fec12a4d07e6359e2fe3d8d52f09a44ace9

Comment (Florian14) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt:181]: Also I think the tombstone event can be redacted, in this case we should probably consider it? It can be done later as this is not a regression and common usecase.

**Comment by Florian14** [comment]
@bmarty Since the sticker event is now also pruned, I observe in the log some exceptions (not crashing) when moshi tries to provide the model of the event content. The expected body should be not null but it became null after the redaction. This is also not a regression because it was already visible after deleting a sticker and doing a clear cache but it's more visible since the event is directly pruned. Wdyt about it? Should we add a default value to the body?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt (hunk 4 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/prune/RedactionEventProcessor.kt:181]: > I removed some to-device event types from the list, tell me if I am wrong [f59be73](https://github.com/vector-im/element-android/commit/f59be7334cd66f056dd791b2ee44a046257b5f9d)

This is LGTM, thanks!

**Comment by bmarty** [comment]
> @bmarty Since the sticker event is now also pruned, I observe in the log some exceptions (not crashing) when moshi tries to provide the model of the event content. The expected body should be not null but it became null after the redaction. This is also not a regression because it was already visible after deleting a sticker and doing a clear cache but it's more visible since the event is directly pruned. Wdyt about it? Should we add a default value to the body?

Yes, maybe just add a default value to avoid such logs at https://github.com/vector-im/element-android/blob/4c95aafe1037819c3ef2afd8e5e9534ebb089284/matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/model/message/MessageStickerContent.kt#L37:

```kotlin
@Json(name = "body") override val body: String = "",
```

Thanks!

**Review by bmarty — APPROVED** [review]
LGTM, thanks!
